Fixing some inconsistencies in the x86 HWIntrinsic APIs

tannergooding opened the pull request:
1. This resolves https://github.com/dotnet/corefx/issues/26433 by adding the missing Sse2.MoveMask API.

2. This resolves a few issues in the Sse2 and Sse41 intrinsics where the signatures differed from the C++ signatures. This also makes the corresponding Sse intrinsics deviate from their C++ signatures to be inline with the rest of the scalar intrinsics.

3. This fixes naming of arguments in LoadHigh and LoadLow to be consistent with the naming of other parameters where the operand is used for the upper or lower bits of the returned result.

tannergooding commented:
I discussed 2 with @fiigii at some length.

My reasonings for the change are basically consistency, clarity, and simplicity:

1. It makes the APIs consistent

We basically have two types of scalar intrinsics. Those that zero the upper bits, and those that set it to the value of one of the source registers. The behavior of these intrinsics matches the behavior of the underlying hardware instructions (some zero the upper bits and some set it to a value of one of the sources).
All of the binary scalar intrinsics (such as `addss`, `divss`, `mulss`, and `subss`) set the value of the upper bits to the first source (left).
The unary scalar intrinsics have a split:
* Those that take a primitive scalar will zero the upper bits (this includes intrinsics like `LoadScalar`, `SetScalar`, `ConvertToVector128Int32Scalar`, etc.)
* Those that take a simd scalar will set the upper bits to the upper bits of one of the source registers

When matching the signature of the C++ intrinsics, all scalar intrinsics except for sqrtss, rsqrtss, and rcpss explicitly take two parameters (a and b) and set the upper bits of the result to the upper bits of a and perform the scalar operation on the lower bits of b. Not having the intrinsics match for the three mentioned meant the user had to remember that the upper bits aren't set to zero for these in particular.

2. It makes the intent of the API clear

For the scalar intrinsics which do set the upper bits, having an explicit upper value makes the result of the operation clear. That is `Sqrt(value)` doesn't clearly tell you what the value of the upper bits will be, and the user has to ensure they read the documentation to determine the precise behavior. `Sqrt(upper, value)` on the other hand makes it very clear what the upper bits will be and what the value that will be operated on will be.

3. It makes the JIT support for these simpler

Not having an explicit parameter also means that the handling for those intrinsics needs their own code path for codegen (they are 1-argument intrinsics that call `SIMD_R_R_R`) and they cannot undergo easily undergo containment (only the second operand can be contained, but both src1 and src2 are op1).

Additionally, it means the user has to manually write additional code to handle the scenario where they want upper bits to differ from the existing upper bits in `value`. The JIT would need to recognize and support this pattern, or the user would be stuck with larger (by number of bytes) and potentially slower code (more instructions filling the pipeline).

tannergooding commented:
FYI. @CarolEidt, @fiigii, @eerhardt

fiigii commented on src/mscorlib/src/System/Runtime/Intrinsics/X86/Sse2.cs:
```diff
@@ -1101,7 +1105,7 @@ public static class Sse2
         /// __m128d _mm_sqrt_sd (__m128d a)
         /// </summary>
-        public static Vector128<double> SqrtScalar(Vector128<double> value) => SqrtScalar(value);
+        public static Vector128<double> SqrtScalar(Vector128<double> upper, Vector128<double> value) => SqrtScalar(upper, value);
```
From the performance perspective, we always want to duplicate the source register for some of these instructions with VEX-encoding. For example, `vsqrtss xmm0, xmm1, xmm1` is better than  `vsqrtss xmm0, xmm0, xmm1` and `vsqrtss xmm0, xmm2, xmm1`. So `SqrtScalar(upper, value)` may make performance issues that is hard to detected.

Form the user experience perspective, `SqrtScalar(value)` usually is more straightforward and useful than `SqrtScalar(upper, value)`.

In summary, I think we can have  `SqrtScalar(value)` and `SqrtScalar(upper, value)` both if we really need the 2-arg version.

tannergooding replied:
> we always want to duplicate the source register for some of these instructions with VEX-encoding.

I think this should say: `The recommendation is to duplicate the source register for some of these instructions.`

I think its one of those things where an analyzer comes in handy (just like the user could do `SetAll(0.0f)`, instead of `SetZero()`, even though the latter is more performant).

>  For example, vsqrtss xmm0, xmm1, xmm1 is better than vsqrtss xmm0, xmm0, xmm1 and vsqrtss xmm0, xmm2, xmm1

However, `vsqrtss target, upper, value` (when upperReg != valueReg) will produce smaller codegen then `movss value, upper, value; vsqrtss target, value, value`, which is what users would need to code otherwise (via `value = MoveScalar(upper, value); target = SqrtScalar(value)`). It is also (theoretically) faster since there are fewer instructions filling the pipeline (the equivalent non-vex instructions behave similarly).

> In summary, I think we can have SqrtScalar(value) and SqrtScalar(upper, value) both if we really need the 2-arg version.

I was in agreement that this could/should be left open as an option. It is not possible to do for the binary scalar simd instructions (like `AddScalar`), but does allow easy exposure of the "recommended" use for the unary scalar simd instructions (with the caveats that it "may" reduce consistency/clarity of the instructions behavior).

4creators replied:
I agree with @fiigii that good solution would be to have to scalar overloads (i) two arg version and (ii) one arg version

fiigii replied:
> It is not possible to do for the binary scalar simd instructions (like AddScalar), 

Absolutely, I meant just for certain 1-arg SIMD scalar intrinsic, such as `SqrtScalar`, `RoundCurrentDirectionScalar`, `ReciprocalScalar`, etc.

CarolEidt replied:
I'm fine with the two overloads, though I begin to be a bit concerned about even further exploding the API surface.

tannergooding replied:
>  though I begin to be a bit concerned about even further exploding the API surface.

That is one of my concerns as well.

These could possibly be implemented in terms of `SqrtScalar(value) => SqrtScalar(value, value)`, with an aggressive inlining attribute or as a general helper library (CoreFXExtensions) for commonly used algorithms/helpers...

fiigii replied:
> These could possibly be implemented in terms of SqrtScalar(value) => SqrtScalar(value, value), with an aggressive inlining attribute or as a general helper library (CoreFXExtensions) for commonly used algorithms/helpers...

That will make JIT/VM more complex because all the methods in this namespace are automatically labeled with [Intrinsic]. If we want the two versions both, implementing them in JIT via the table-driven framework is really easy.

4creators replied:
> a bit concerned about even further exploding the API surface

We are much more concise than C++ implementation. The overloads which are proposed are the closest match to original C++ intrinsics and there will be only few of them. This should not have any significant impact on number of methods which would be around 1 300 for current accepted Intel ISA. I have not counted Arm but it will be significantly more concise as there are more methods which can be expressed as generics according to our API design rules.

tannergooding replied:
AH, I forgot about the implicit `[Intrinsic]` handling for all methods here

CarolEidt commented on src/mscorlib/src/System/Runtime/Intrinsics/X86/Sse.cs:
```diff
@@ -344,7 +344,7 @@ public static class Sse
         /// __m128 _mm_rcp_ss (__m128 a)
         /// </summary>
-        public static Vector128<float> ReciprocalScalar(Vector128<float> value) => ReciprocalScalar(value);
+        public static Vector128<float> ReciprocalScalar(Vector128<float> upper, Vector128<float> value) => ReciprocalScalar(upper, value);
```
This leads to confusion where the C++ intrinsic doesn't include the additional argument, and makes me wonder whether we should include *both* the C++intrinsic as well as the actual opcode, and use the parameter names, e.g. for this case we would have:
```
         /// <summary>
          /// __m128 _mm_rcp_ss (__m128 value)
          /// rcpps upper, value
          /// </summary>
```
Thoughts?

tannergooding replied:
There are only 3 intrinsics which differ (`rcp_ss`, `rsqrt_ss`, and `sqrt_ss`), we could probably just add an explicit comment.
```
/// <summary>
/// __m128 _mm_rcp_ss (__m128 value)
/// We deviate from the native signature and allow the user to specify upper, rather than having it implicitly come from value. This is for consistency with the other scalar intrinsics.
/// </summary>
```

fiigii commented:
> These could possibly be implemented in terms of SqrtScalar(value) => SqrtScalar(value, value), with an aggressive inlining attribute or as a general helper library (CoreFXExtensions) for commonly used algorithms/helpers...

That will make JIT/VM more complex because all the methods in this namespace are automatically labeled with `[Intrinsic]`. If we want the two versions both, implementing them in JIT via the table-driven framework is really easy.

tannergooding commented:
Addressed PR feedback.

For the SSE2 and SSE4.1 intrinsics, I added an overload which only takes value. These also have a comment stating: `The above native signature does not exist. We provide this additional overload for the recommended use case of this intrinsic.`

For the SSE intrinsics, I added an overload which takes upper and value. These also have a comment stating: `The above native signature does not exist. We provide this additional overload for consistency with the other scalar APIs.`

This adds a total of 14 new APIs.

4creators commented:
Actually there a couple of other missing APIs in Sse2 - see https://github.com/dotnet/corefx/issues/25926#issuecomment-359447469. These are:

```C#
// __m128i _mm_move_epi64 (__m128i a)
public static Vector128<long> MoveScalar(Vector128<long> upper, Vector128<long> value) => MoveScalar(upper, value);
public static Vector128<ulong> MoveScalar(Vector128<ulong> upper, Vector128<ulong> value) => MoveScalar(upper, value);

// __m64 _mm_movepi64_pi64 (__m128i a)
public static long* StoreScalar(Vector128<long> value);
public static ulong* StoreScalar(Vector128<ulong> value);

// void _mm_clflush (void const* p)
public static void FlushCache(void* p);

// void _mm_mfence (void)
public static void MemoryFence();
```

fiigii commented:
There is one more inconsistency:
```csharp
// SSE4.1
public static unsafe Vector128<sbyte> LoadAlignedNonTemporal(sbyte* address) {}

// AVX2
public static unsafe Vector256<sbyte> LoadAlignedVector256NonTemporal(sbyte* address) {}
```

@tannergooding could you helpe me to change `LoadAlignedNonTemporal` to `LoadAlignedVector128NonTemporal` in this PR?

Sorry for the inconvenience.

tannergooding commented:
Updated. Adding documentation on what instruction an intrinsic maps to.

The below list shows the intrinsics that are not exposed.
* MMX instructions aren't exposed because they generally aren't useful and cause a transition penalty
* Explicitly not exposed instructions are generally special memory fencing instructions which was decided to be reviewed at a later date/time
* Partially exposed instructions are just the reg/reg encodings for the move instructions.

* SSE
  * MMX Extensions (cause transition from x87 FPU to MMX)
    * CVTPI2PS
    * CVTPS2PI
    * CVTTPS2PI
    * MASKMOVQ
    * PAVGB
    * PAVGW
    * PEXTRW
    * PINSWR
    * PMAXUB
    * PMAXSW
    * PMINUB
    * PMINSW
    * PMOVMSKB
    * PMULHUW
    * PSADBW
    * PSHUFW
  * MMX Extensions (transition not listed)
    * MOVNTQ
  * Not Exposed (explicit)
    * LDMXCSR
    * PREFETCHh
    * SFENCE
    * STMXCSR
  * Partially Exposed (missing encodings listed)
    * MOVAPS xmm, xmm
    * MOVUPS xmm, xmm
* SSE2
  * MMX Extensions (cause transition from x87 FPU to MMX)
    * CVTPI2PD
    * CVTPD2PI
    * CVTTPD2PI
    * MOVQ2DQ
    * MOVDQ2Q
  * Not Exposed (explicit)
    * CLFLUSH
    * LFENCE
    * MFENCE
    * PAUSE
  * Not Exposed (unknown)
    * MOVNTI
  * Partially Exposed (missing encodings listed)
    * MOVAPD xmm, xmm
    * MOVUPD xmm, xmm
* SSE3
  * x87 FPU Extensions
    * FISTTP
  * Not Exposed (explicit)
    * MONITOR
    * MWAIT
* SSE4.1
  * Not Exposed (unknown)
    * PEXTRW
* AVX
  * Partially Exposed (missing encodings listed)
    * VMOVDQU ymm, ymm
    * VMOVUPS ymm, ymm
    * VMOVUPD ymm, ymm
    * VMOVDQA ymm, ymm
    * VMOVAPS ymm, ymm
    * VMOVAPD ymm, ymm

4creators commented:
SSE4.1 `PEXTRW reg/m16, xmm, imm8` allows to extract directly to memorywhat is not possible with SSE2 instructions. All other `PEXTRB/D/Q` allow to extract to memory or register. This last instruction unifies all extract API. Perhaps we should include it as well.

tannergooding commented:
@fiigii, thoughts?

fiigii commented on src/mscorlib/src/System/Runtime/Intrinsics/X86/Fma.PlatformNotSupported.cs:
```diff
@@ -15,140 +15,140 @@ public static class Fma
         public static Vector128<float> MultiplyAdd(Vector128<float> a, Vector128<float> b, Vector128<float> c) { throw new PlatformNotSupportedException(); }
         /// <summary>
-        /// __m128d _mm_fmadd_pd (__m128d a, __m128d b, __m128d c)
+        /// __m128d _mm_fmadd_pd (__m128d a, __m128d b, __m128d c); VFMADDPD xmm, xmm, xmm/m128
```
Putting the instruction in a new line maybe looks better?
```csharp
/// <summary>
/// __m128d _mm_fmadd_pd (__m128d a, __m128d b, __m128d c)
/// VFMADDPD xmm, xmm, xmm/m128
/// </summary>
```

tannergooding replied:
Possibly. It doesn't look as good for the intrinsics with multiple mappings, however.

The current format will work fairly well for a search and replace to improve the documentation however.

fiigii replied:
@eerhardt could you suggest the comment format?

CarolEidt replied:
FWIW, my preference would also be to have the instruction(s) on a separate line.

eerhardt replied:
I agree that separate lines are more readable.

tannergooding replied:
Fixed

tannergooding commented on src/mscorlib/src/System/Runtime/Intrinsics/X86/Avx.cs:
```diff
@@ -14,223 +14,223 @@ namespace System.Runtime.Intrinsics.X86
+        /// __m256d _mm256_cvtps_pd (__m128 a); VCVTPS2PD ymm, xmm/m128
         /// </summary>
-        public static Vector256<double> ConvertToVector256Double(Vector256<float> value) => ConvertToVector256Double(value);
+        public static Vector256<double> ConvertToVector256Double(Vector128<float> value) => ConvertToVector256Double(value);
```
The parameter was incorrect, changed from `Vector256` to `Vector128`

fiigii replied:
Thank you!!

tannergooding commented on src/mscorlib/src/System/Runtime/Intrinsics/X86/Avx.cs:
```diff
@@ -14,223 +14,223 @@ namespace System.Runtime.Intrinsics.X86
+        /// __m256d _mm256_cvtepi32_pd (__m128i a); VCVTDQ2PD ymm, xmm/m128
         /// </summary>
-        public static Vector256<double> ConvertToVector256Double(Vector256<int> value) => ConvertToVector256Double(value);
+        public static Vector256<double> ConvertToVector256Double(Vector128<int> value) => ConvertToVector256Double(value);
```
The parameter was incorrect, changed from `Vector256` to `Vector128`

tannergooding commented on src/mscorlib/src/System/Runtime/Intrinsics/X86/Avx2.cs:
```diff
@@ -221,1293 +221,1293 @@ public static class Avx2
-        /// __m256i _mm256_max_epu8 (__m256i a, __m256i b)
+        /// __m256i _mm256_max_epu8 (__m256i a, __m256i b); VPMAXUB ymm, ymm, ymm/m256
         /// </summary>
-        public static Vector256<byte> Max(Vector256<byte> left, Vector256<short> right) => Max(left, right);
```
Second parameter's base type was incorrect. Changed from `short` to `byte`

fiigii commented:
> @fiigii, thoughts?

Do you mean the unexposed instructions? I think we talked most of them previously. 
For `SSE4.1 PEXTRW reg/m16, xmm, imm8`, I think we should follow the current design that does not expose the encoding form as intrinsics. This instruction should be generated by folding `Store*`.

tannergooding commented on src/mscorlib/src/System/Runtime/Intrinsics/X86/Sse.cs:
```diff
@@ -16,373 +16,386 @@ public static class Sse
-        /// __m128 _mm_load_ss (float const* mem_address)
+        /// __m128 _mm_load_ss (float const* mem_address); MOVSS xmm, m32
         /// </summary>
-        public static unsafe Vector128<float> LoadScalar(float* address) => LoadScalar(address);
```
Fixed name to be `LoadScalarVector128`

fiigii replied:
Not necessary, we do not have `LoadScalarVector256`, and `Scalar` already implies `Vector128`.

tannergooding replied:
I changed it here for consistency with the other APIs naming convention. I figured it was better to be consistent.

fiigii replied:
Fair enough.

CarolEidt replied:
I also prefer the addition of `Vector128` to the name; I think it makes it clearer, especially since when encountered in code it may not always be clear that it is returning a vector.

tannergooding commented on src/mscorlib/src/System/Runtime/Intrinsics/X86/Sse.cs:
```diff
@@ -16,373 +16,386 @@ public static class Sse
-        /// __m128 _mm_loadh_pi (__m128 a, __m64 const* mem_addr)
+        /// __m128 _mm_loadh_pi (__m128 a, __m64 const* mem_addr); MOVHPS xmm, m64
         /// </summary>
-        public static unsafe Vector128<float> LoadHigh(Vector128<float> value, float* address) => LoadHigh(value, address);
```
Name of first parameter change to be `lower`

tannergooding commented on src/mscorlib/src/System/Runtime/Intrinsics/X86/Sse.cs:
```diff
@@ -16,373 +16,386 @@ public static class Sse
-        /// __m128 _mm_loadl_pi (__m128 a, __m64 const* mem_addr)
+        /// __m128 _mm_loadl_pi (__m128 a, __m64 const* mem_addr); MOVLPS xmm, m64
         /// </summary>
-        public static unsafe Vector128<float> LoadLow(Vector128<float> value, float* address) => LoadLow(value, address);
```
Name of first parameter changed to be `upper`

tannergooding commented on src/mscorlib/src/System/Runtime/Intrinsics/X86/Sse.cs:
```diff
@@ -16,373 +16,386 @@ public static class Sse
-        /// __m128 _mm_set_ss (float a)
+        /// __m128 _mm_set_ss (float a); HELPER
         /// </summary>
-        public static Vector128<float> SetScalar(float value) => SetScalar(value);
```
Name changed to be `SetScalarVector128`

fiigii replied:
Not necessary,

tannergooding commented on src/mscorlib/src/System/Runtime/Intrinsics/X86/Sse2.cs:
```diff
@@ -14,858 +14,870 @@ namespace System.Runtime.Intrinsics.X86
-        /// __m128d _mm_loadh_pd (__m128d a, double const* mem_addr)
+        /// __m128d _mm_loadh_pd (__m128d a, double const* mem_addr); MOVHPD xmm, m64
         /// </summary>
-        public static unsafe Vector128<double> LoadHigh(Vector128<double> value, double* address) => LoadHigh(value, address);
```
Name of first parameter changed to `lower`.

tannergooding commented on src/mscorlib/src/System/Runtime/Intrinsics/X86/Sse2.cs:
```diff
@@ -14,858 +14,870 @@ namespace System.Runtime.Intrinsics.X86
-        /// __m128d _mm_loadl_pd (__m128d a, double const* mem_addr)
+        /// __m128d _mm_loadl_pd (__m128d a, double const* mem_addr); MOVLPD xmm, m64
         /// </summary>
-        public static unsafe Vector128<double> LoadLow(Vector128<double> value, double* address) => LoadLow(value, address);
```
Name of first parameter changed to `upper`.

tannergooding commented on src/mscorlib/src/System/Runtime/Intrinsics/X86/Sse2.cs:
```diff
@@ -14,858 +14,870 @@ namespace System.Runtime.Intrinsics.X86
-        /// __m128d _mm_set_sd (double a)
+        /// __m128d _mm_set_sd (double a); HELPER
         /// </summary>
-        public static Vector128<double> SetScalar(double value) => SetScalar(value);
```
Named changed to `SetScalarVector128`

4creators replied:
IMO from user perspective it is better to have names with Vector128/Vector256 even when Vector128 is implied.

fiigii commented:
Thank you so much for the work!

fiigii commented on src/mscorlib/src/System/Runtime/Intrinsics/X86/Sse2.cs:
```diff
@@ -874,536 +886,542 @@ public static class Sse2
         public static Vector128<ushort> UnpackHigh(Vector128<ushort> left, Vector128<ushort> right) => UnpackHigh(left, right);
         /// <summary>
-        /// __m128i _mm_unpackhi_epi32 (__m128i a,  __m128i b)
+        /// __m128i _mm_unpackhi_epi32 (__m128i a,  __m128i b); PUNPCKHDQ xmm, xmm/m128
```
Could you fix the above `/// </summary`?

fiigii commented on src/mscorlib/src/System/Runtime/Intrinsics/X86/Sse2.PlatformNotSupported.cs:
```diff
@@ -15,1391 +15,1410 @@ public static class Sse2
         /// </summary>
-        public static Vector128<byte> Or(Vector128<byte> left,  Vector128<byte> right) { throw new PlatformNotSupportedException(); }
+        public static Vector128<byte> Or(Vector128<byte> left, Vector128<byte> right) { throw new PlatformNotSupportedException(); }
       
```
And here.

fiigii commented on src/mscorlib/src/System/Runtime/Intrinsics/X86/Avx2.PlatformNotSupported.cs:
```diff
@@ -14,1491 +14,1492 @@ namespace System.Runtime.Intrinsics.X86
-        /// __m256i _mm256_inserti128_si256 (__m256i a, __m128i b, const int imm8)
+        /// __m256i _mm256_inserti128_si256 (__m256i a, __m128i b, const int imm8); VINSERTI128 ymm, ymm, m128, imm8
         /// </summary>
         public static unsafe Vector256<long> Insert(Vector256<long> value, long* address,
```
Thanks!

CarolEidt commented on src/mscorlib/src/System/Runtime/Intrinsics/X86/Sse2.cs:
```diff
@@ -14,858 +14,870 @@ namespace System.Runtime.Intrinsics.X86
-        /// __m128i _mm_mul_epu32 (__m128i a,  __m128i b)
+        /// __m128i _mm_move_epi64 (__m128i a); MOVQ xmm, xmm
+        /// </summary>
+        public static Vector128<long> MoveScalar(Vector128<long> value) => MoveScalar(value);
```
What is the reasoning for this being `MoveScalar`, but changing `SetScalar` to `SetScalarVector128` when I believe in both cases the entire 128-bit vector is defined.

tannergooding replied:
The former (`MoveScalar`) takes and returns `Vector128`, so it shouldn't cause overload resolution problems. The latter (`SetScalar`) takes a primitive and returns a `Vector128`, so it could potentially conflict.

I can change if you think its clearer to provide it here as well.

CarolEidt replied:
I think it would be more consistent, but I don't feel strongly about it.

CarolEidt commented on src/mscorlib/src/System/Runtime/Intrinsics/X86/Sse41.PlatformNotSupported.cs:
```diff
@@ -14,387 +14,459 @@ namespace System.Runtime.Intrinsics.X86
 
         /// <summary>
-        /// __m128 _mm_ceil_ps (__m128 a)
+        /// __m128 _mm_ceil_ps (__m128 a); ROUNDPS xmm, xmm/m128, imm8(10)
```
Would it be clearer to write the implicit immediate as `imm8=0b10`?

tannergooding replied:
Probably. 👍

fiigii commented:
@tannergooding @CarolEidt @mikedn Shall we change the APIs of `LoadScalar` to take `ref` parameter instead of pointer?
That keeps consistent with `BroadcastElementToVector128/256`, so that load/store from/to a scalar value via `ref` because it does not need alignment.

mikedn commented:
> Shall we change the APIs of LoadScalar to take ref parameter instead of pointer?

Sounds like a good idea to me, pointers requires unsafe code but there's nothing unsafe here. But then, I haven't paid too much attention to the issues that added these APIs so I'm not sure how we ended up with pointers. I'm not even sure why we have stuff like Load(pointer).

CarolEidt commented:
> Shall we change the APIs of LoadScalar to take ref parameter instead of pointer?

I think so

tannergooding commented:
> I'm not sure how we ended up with pointers. I'm not even sure why we have stuff like Load(pointer).

Using refs was discussed, but it wasn't clear if it would make sense for `ref float` to actually be `ref float[4]`. It is also less than easy (but only slightly) to convert between `ref T` and `ref U`, while it is trivial to do so for `T*` and `U*`.

All of the expected interchange formats (Span, array, etc) also provide an easy way to get a pointer (without having to pull in a third party dependency, etc).

Additionally, the C# `unmanaged` constraint will allow pointers to work with some generics in some future release: https://github.com/dotnet/csharplang/pull/1276

tannergooding commented:
> Shall we change the APIs of LoadScalar to take ref parameter instead of pointer?

@fiigii, @CarolEidt.

Are we sure we want to do that? It forces a dependency on `System.Runtime.CompilerServices.Unsafe` for the cases where the user needs to call `LoadScalar` (or `BroadcastElementToVector`) but when they don't have a `float` (if they have a `byte[]`, for example; then they must call `ref Unsafe.As<byte, float>(ref x[i])`).

mikedn commented:
> but when they don't have a float (if they have a byte[], for example; then they must call ref Unsafe.As<byte, float>(ref x[i])).

I don't know but I'd guess that if someone wants to load a `float` from a `byte[]` like that then it's not unreasonable to expect that they'd also do other weird things that require `Unsafe` anyway. It is what it is, guesswork. A large API is being added and it's not always clear what exactly will be needed.

tannergooding commented:
>  that they'd also do other weird things that require Unsafe anyway.

Agreed. My point, however, is that if they are having to do unsafe things anyways, pointers are a bit more clear, don't require another NuGet dependency, and ensures the user has thought about stuff like pinning/alignment.

Working with `ref` is nice, but when you have to treat it like a pointer (dealing with offsets, conversions, etc), it makes the code much more verbose and a bit more painful to work with. Ex: What could be `x += offset` or `(float*)(x)`, becomes `x = ref Unsafe.Add(ref x, offset)` and `ref Unsafe.As<byte, float>(ref x)`.

tannergooding commented:
I don't have a strong preference, and am happy to change to ref if everyone agrees, but my preference is on using pointers over ref.

CarolEidt commented:
> my preference is on using pointers over ref.

That sounds like something we should reach a consensus on and make some general guidelines. I can see value on both sides but lean toward ref.

4creators commented:
AFAIR the types used in interchange APIs were **very hotly debated** before API review and arriving finally at pointers was deemed as a best and most flexible choice. Therefore, if we drop pointers we loose a lot of flexibility and low level control which comes with them.

However, there are very strong arguments in favor of ref parameters and one of the most important was an ability to write generic algorithms and have good fit with `Span<T>` based APIs. see https://github.com/dotnet/corefx/issues/22940#issuecomment-320286336 and https://github.com/dotnet/corefx/issues/22940#issuecomment-320292507

IMO it would be best to have overloads for both pointers and ref parameters if we do not want to have very disappointed developers.

see https://github.com/dotnet/corefx/issues/22940#issuecomment-320230371 and many many more posts below

tannergooding commented:
There are some issues with pointers:
* Not all languages can use them
* They don't play well with generics
  * This is being resolved by the C# `unmanaged` constraint

There are some issues with `ref` as well:
* Not all languages can use them
* It is hard to interact with a ref as if it was a pointer

4creators commented:
> There are some issues with pointers
> There are some issues with ref as well

This is why having both of them is the best solution for all - number of interchange APIs is not large and adding overloads will not cause significant API increase.

fiigii commented:
I remember that the initial motivition of introducing these `*Scalar` intrinsics is implementing `Math` lib in managed code. It may be useful to investigate if we indeed want the code like `ref Unsafe.As<byte, float>(ref x[i]))` in managed `Math`. Meanwhile, I think users won't need to use `*Scalar` intrinsics so much, for these instructions are already covered by RyuJIT codgen for FP.

> I can see value on both sides but lean toward ref

Me too.

tannergooding commented:
>> I can see value on both sides but lean toward ref
>
>Me too.

Wondering what the motivation here is? Is it just because it isn't "unsafe", or is there other reasons as well?

fiigii commented:
>  Is it just because it isn't "unsafe"

Yes, I guess `ref` may be more convenient to use than unsafe because `Scalar` does not require alignment. However, as I said above, if managed `Math` will need "unsafe", we should provide the pointer version at first.
BTW, `BroadcastElementToVector128/256` should be consistent with `LoadScalar`.

tannergooding commented:
> will need "unsafe", we should provide the pointer version at first.

I don't think there is any API which will have to have one vs the other. Both are equally as flexibile in what they can do:
* You can pin a ref, offset it, convert it from T to U, convert it to a pointer, etc
* You can also do all the same things with pointers

The caveat on the former is that `ref` requires the use of `S.R.CS.Unsafe` in order to perform many of the operations.

The caveat on the latter is that you have to enable the unsafe flag/put a block of code in an "unsafe" region and it won't work with generics until the next C# langver.

In both cases, you can argue that the user is already doing things which would require the `Unsafe` library or the `unsafe` compiler flag.

My preference is on using pointers because it is considerably less verbose and (IMO) is slightly easier to read.

tannergooding commented:
Updated the post above, forgot to list that we are missing several `Avx2.BroadcastElementToVector` methods (specifically, the ones that take `ref T`).

@fiigii, could you comment on the missing and partially exposed instructions from the list [above](https://github.com/dotnet/coreclr/pull/15923#issuecomment-360266341), on whether you think they were missed or why they were excluded (I already listed the exclusion reason on a few that I knew were purposefully excluded)?

fiigii commented:
> AVX2
Partially Exposed (missing encodings listed)
VPBROADCASTB xmm, m8
VPBROADCASTB ymm, m8


I think these encodings should be generated by containment analysis (folding Load*) on `BroadcastElementToVector128/256<T>` and `SetAllVector128/256<T>`.


> SSE4.1
Not Exposed (unknown)
PEXTRW


This instruction also should be generated by containment analysis (folding Store*).
The current API design of Intel HW intrinsic does not expose specific encoding as intrinsic except common programming patterns require them.


> Partially Exposed (missing encodings listed)
MOVAPD xmm, xmm
Partially Exposed (missing encodings listed)
VMOVDQU ymm, ymm
VMOVUPS ymm, ymm
VMOVUPD ymm, ymm

The current RyuJIT already has a good general block-copy, which always uses `MOVAPS` to move data between registers (shorter encoding and zero-latency)

CarolEidt commented:
> AFAIR the types used in interchange APIs were very hotly debated before API review and arriving finally at pointers was deemed as a best and most flexible choice. Therefore, if we drop pointers we loose a lot of flexibility and low level control which comes with them.

> However, there are very strong arguments in favor of ref parameters and one of the most important was an ability to write generic algorithms and have good fit with Span<T> based APIs.

Could we get some input on these points from API designers - @eerhardt @terrajobst ?

fiigii commented:
> PREFETCHh

Usually, hardware prefetch should be sufficient, and overuse of software prefetch is discouraged. However, I have seen some demand for `PREFETCH` intrinsics. Probably, we should expose this intrinsics.

tannergooding commented:
> I think these encodings should be generated by containment analysis (folding Load*)

We are currently exposing these encodings in `AVX` for `float` and `double`, should those encodings be removed (for consistency)?

tannergooding commented:
> Usually, hardware prefetch should be sufficient, and overuse of software prefetch is discouraged

I think it would be good to review the various prefetch and fencing instructions in bulk at a later date (looks like there are 8 of them, in total).

fiigii commented:
> We are currently exposing these encodings in AVX for float and double

Don't you mean `BroadcastElementToVector*`? `vbroadcastss` and `vbroadcastsd` only have `xmm/ymm, m32/m64` encoding in AVX.

tannergooding commented:
> Don't you mean BroadcastElementToVector*?

Hmmm, I guess for `VBROADCAST`, it is:
* VBROADCASTSS xmm1, m32: `BroadcastElementToVector128(ref float source)`
* VBROADCASTSS ymm1, m32: `BroadcastElementToVector256(ref float source)`
* VBROADCASTSD ymm1, m64: `BroadcastElementToVector256(ref double source)`
* VBROADCASTF128 ymm1, m128: `BroadcastVector128ToVector256(float* address)` and `BroadcastVector128ToVector256(double* address)`

While for `VPBROADCAST`, it is:
* VPBROADCASTB xmm1, xmm2/m8: `BroadcastElementToVector128<T>(Vector128<T> value)`
* VPBROADCASTB ymm1, xmm2/m8: `BroadcastElementToVector256<T>(Vector128<T> value)`
* VPBROADCASTW xmm1, xmm2/m16: `BroadcastElementToVector128<T>(Vector128<T> value)`
* VPBROADCASTW ymm1, xmm2/m16: `BroadcastElementToVector256<T>(Vector128<T> value)`
* VPBROADCASTD xmm1, xmm2/m32: `BroadcastElementToVector128<T>(Vector128<T> value)`
* VPBROADCASTD ymm1, xmm2/m32: `BroadcastElementToVector256<T>(Vector128<T> value`
* VPBROADCASTQ xmm1, xmm2/m64: `BroadcastElementToVector128<T>(Vector128<T> value)`
* VPBROADCASTQ ymm1, xmm2/m64: `BroadcastElementToVector256<T>(Vector128<T> value)`
* VBROADCASTI128 ymm1, m128: `BroadcastVector128ToVector256(T* address)`

It's initially confusing that we are exposing `(ref T source)` for one, but not the other.

It also looks like we are exposing:
* VBROADCASTSS ymm, xmm: `BroadcastElementToVector128<T>(Vector128<T> value)`
* VBROADCASTSD ymm, xmm: `BroadcastElementToVector128<T>(Vector128<T> value)`

But I don't actually see a `VBROADCASTSS ymm, xmm` or `VBROADCASTSD ymm, xmm` encoding in the Architecture Software Developer's Manual.

fiigii commented:
@jkotas Do you have suggestions for exposing `PREFETCH` intrinsics or not?

jkotas commented:
I do not see a problem with. It is a drop in a bucket given the total number of these APIs.

eerhardt commented:
>>I'm not sure how we ended up with pointers. I'm not even sure why we have stuff like Load(pointer).

> Using refs was discussed, but it wasn't clear if it would make sense for ref float to actually be ref float[4].

I think this comment explains why the Load APIs take pointers when dealing with non-scalars.  Using `ref float` to actually mean more than just that float 32-bit location is wrong and why we didn't go with it.

> BTW, BroadcastElementToVector128/256 should be consistent with LoadScalar.

Agreed.  When we have an input scalar parameter, we should be consistent across the APIs.  
The same goes for when we have an output scalar parameter.  One output scalar parameter shouldn't use `ref` or `out`, while a similar API use a pointer.

And then the next level of consistency would be between input and output scalar parameters.  If input scalar parameters use `ref` or `in`, then it follows that we should have output scalar parameters use `ref` or `out`.  Similarly, if output parameters use pointers, I think input parameters should use pointers as well.

It seems to me that pointers will work, especially if and when the C# generics support is implemented.  If we feel we also need `in` and `out` overloads, I think we can add them (either now, or later when there is a hard requirement for them).

jkotas commented:
> It seems to me that pointers will work,

Yes, we should be using pointers only in these APIs.

CarolEidt commented:
Thanks @eerhardt and @jkotas. It looks like we have a guideline.

@tannergooding - there have been some suggestions wrt the comment formats (instruction on separate line, and using the C# constant syntax, e.g. `0b10` for binary 10) and adding the `Vector128` to some names. Are you planning to make those changes?

tannergooding commented:
> Are you planning to make those changes?

Yes, I should get these changes up tonight.

fiigii commented:
> But I don't actually see a VBROADCASTSS ymm, xmm or VBROADCASTSD ymm, xmm encoding in the Architecture Software Developer's Manual. 

Yes, this was a mistake in the older version of the manual. The [latest manual](https://software.intel.com/sites/default/files/managed/39/c5/325462-sdm-vol-1-2abcd-3abcd.pdf) has fixed it (See page 5-12). 

`VBROADCASTSS/D` only has `reg, mem` encoding in AVX that we expose as "load with broadcast".
AVX2 complemented the `reg, reg` encdoing, so that we can have `BroadcastElementToVector128/256<T>(Vector128<T> value)`.

tannergooding commented:
Ok. I think all questions/feedback have been answered/resolved.

We are using pointers everywhere
The associated instruction is now on its own line
We are using  VectorXX in the name to disambiguate anywhere where said Vector is returned, but is not one of the arguments passed in.
The missing intrinsics that are not being handled by containment are all exposed.

tannergooding commented:
@CarolEidt, @fiigii, @eerhardt. I think this should be ready for a final review now.

4creators commented on src/mscorlib/src/System/Runtime/Intrinsics/X86/Avx2.cs:
```diff
@@ -221,1293 +275,1597 @@ public static class Avx2
         public static Vector256<ulong> ConvertToVector256UInt64(Vector128<byte> value) => ConvertToVector256UInt64(value);
         /// <summary>
-        /// __m256i _mm256_cvtepu16_epi64 (__m128i a)
+        /// __m256i _mm256_cvtepu16_epi64 (__m128i a; VPMOVZXWQ ymm, xmm/m128)
```
nit: instruction on new line

tannergooding replied:
Fixed.

4creators commented on src/mscorlib/src/System/Runtime/Intrinsics/X86/Bmi2.cs:
```diff
@@ -49,6 +49,6 @@ public static class Bmi2
         /// <summary>
         /// unsigned __int64 _pext_u64 (unsigned __int64 a, unsigned __int64 mask)
         /// </summary>
```
nit: missing instruction

tannergooding replied:
I didn't update any of the BMI, AES, LZCNT, or POPCNT ISAs (looks like this picked up a trailing whitespace removal).

I'll update these real quick.

tannergooding replied:
Fixed.

jkotas commented on src/mscorlib/src/System/Runtime/Intrinsics/X86/Sse.PlatformNotSupported.cs:
```diff
@@ -16,444 +16,589 @@ public static class Sse
+        /// void _mm_prefetch(char* p, int i);
+        ///   PREFETCHT1 m8
+        /// </summary>
+        public static unsafe void Prefetch1(byte* address) { throw new PlatformNotSupportedException(); }
```
Should this rather have `void*` argument?

tannergooding replied:
Probably, will update.

tannergooding replied:
Fixed.

4creators commented on src/mscorlib/src/System/Runtime/Intrinsics/X86/Sse.cs:
```diff
@@ -16,373 +16,490 @@ public static class Sse
+        ///   MOVSS xmm, m32
         /// </summary>
-        public static unsafe Vector128<float> LoadScalar(float* address) => LoadScalar(address);
+        public static unsafe Vector128<float> LoadScalarVector128(float* address) => LoadScalarVector128(address);
```
In other place we use convention `ConvertToVector128Scalar` with Scalar at last place. Perhaps we should unify naming with regard to position of scalar to vector.

tannergooding replied:
@CarolEidt, @fiigii, @eerhardt. Preference on whether we do:
* `LoadVector128Aligned`
* `LoadVector128Scalar`
* `ConvertToVector128Scalar`
* etc

 or
* `LoadAlignedVector128`
* `LoadScalarVector128`
* `ConvertToScalarVector128`

?

tannergooding replied:
Actually, looking at these, we are fairly consistent already (see https://github.com/dotnet/coreclr/pull/15923#issuecomment-361456561).

We are basically always using the form:
* Action
* Type
* Modifier

CarolEidt replied:
> We are basically always using the form:
> * Action
> * Type
> * Modifier

Seems reasonable, and good to be consistent.

4creators commented on src/mscorlib/src/System/Runtime/Intrinsics/X86/Sse.cs:
```diff
@@ -16,373 +16,490 @@ public static class Sse
+        /// __m128 _mm_load_ps (float const* mem_address);
+        ///   MOVAPS xmm, m128
         /// </summary>
         public static unsafe Vector128<float> LoadAlignedVector128(float* address) => LoadAlignedVector128(address);
```
Perhaps the naming could follow the scheme  `action` `on whom` `how` -> `LoadVector128Aligned`

tannergooding replied:
See my comment here: https://github.com/dotnet/coreclr/pull/15923#issuecomment-361456561

I don't think `Aligned` is the how, rather I think `Aligned Vector` is the type.

4creators commented on src/mscorlib/src/System/Runtime/Intrinsics/X86/Sse.cs:
```diff
@@ -16,373 +16,490 @@ public static class Sse
+        /// void _mm_prefetch(char* p, int i);
+        ///   PREFETCHT0 m8
+        /// </summary>
+        public static unsafe void Prefetch0(byte* address) => Prefetch0(address);
```
👍

4creators commented on src/mscorlib/src/System/Runtime/Intrinsics/X86/Sse.cs:
```diff
@@ -16,373 +16,490 @@ public static class Sse
+        /// __m128 _mm_set1_ps (float a);
+        ///   HELPER
         /// </summary>
         public static Vector128<float> SetAllVector128(float value) => SetAllVector128(value);
```
It's not directly related to this commit but I forget to ask it - perhaps we should generate `(V)BROADCAST` instructions when on AVX/AVX2

tannergooding replied:
Yes. Helper instructions don't make a guarantee about a particular instruction being emitted, they actually make a guarantee about behavior instead.

4creators commented on src/mscorlib/src/System/Runtime/Intrinsics/X86/Sse2.cs:
```diff
@@ -14,858 +14,1080 @@ namespace System.Runtime.Intrinsics.X86
+        ///   PMINUB xmm, xmm/m128
         /// </summary>
         public static Vector128<byte> Min(Vector128<byte> left, Vector128<byte> right) => Min(left, right);
         /// <summary>
```
nit: missing empty line

tannergooding replied:
Missing newline where?

4creators replied:
before line 835 - usually we have one empty line between declarations but sometime they are "missing"

tannergooding replied:
There are a lot of intrinsics which currently only have newlines between logical groupings (e.g. after all Min methods, there is a newline, or after all Add methods, there is a newline).

4creators replied:
OK - it is not visible that well in diffs

tannergooding replied:
agreed. But it is the current "convention".

I'd rather fix it up in a separate PR, if we want to do so.

4creators replied:
Yeah, it is small editorial detail and we will have some opportunities to fix it soon

4creators commented on src/mscorlib/src/System/Runtime/Intrinsics/X86/Sse2.cs:
```diff
@@ -14,858 +14,1080 @@ namespace System.Runtime.Intrinsics.X86
+        ///   PMINSW xmm, xmm/m128
         /// </summary>
         public static Vector128<short> Min(Vector128<short> left, Vector128<short> right) => Min(left, right);
         /// <summary>
```
nit: as above

4creators commented on src/mscorlib/src/System/Runtime/Intrinsics/X86/Sse2.cs:
```diff
@@ -14,858 +14,1080 @@ namespace System.Runtime.Intrinsics.X86
+        ///   PAVGW xmm, xmm/m128
         /// </summary>
         public static Vector128<ushort> Average(Vector128<ushort> left, Vector128<ushort> right) => Average(left, right);
 
```
Seems that `void CasheLineFlush(void* pointer)` or `CLFLUSH` is missing

tannergooding replied:
Controlled via the CLFSH CPUID flag, so it would need to be exposed under a separate ISA

4creators replied:
Right

tannergooding replied:
Given that it is a new ISA, we should do it independently.

4creators commented on src/mscorlib/src/System/Runtime/Intrinsics/X86/Sse2.cs:
```diff
@@ -14,858 +14,1080 @@ namespace System.Runtime.Intrinsics.X86
+        ///   MOVSD xmm, m64
         /// </summary>
-        public static unsafe Vector128<double> LoadScalar(double* address) => LoadScalar(address);
+        public static unsafe Vector128<double> LoadScalarVector128(double* address) => LoadScalarVector128(address);
```
naming - as above

4creators commented on src/mscorlib/src/System/Runtime/Intrinsics/X86/Sse2.cs:
```diff
@@ -14,858 +14,1080 @@ namespace System.Runtime.Intrinsics.X86
+        ///   PMOVMSKB reg, xmm
         /// </summary>
         public static int MoveMask(Vector128<sbyte> value) => MoveMask(value);
         /// <summary>
```
nit: line

4creators commented on src/mscorlib/src/System/Runtime/Intrinsics/X86/Sse2.cs:
```diff
@@ -14,858 +14,1080 @@ namespace System.Runtime.Intrinsics.X86
+        ///   PMOVMSKB reg, xmm
+        /// </summary>
+        public static int MoveMask(Vector128<byte> value) => MoveMask(value);
+        /// <summary>
```
nit: I will stop commenting on missing empty lines

4creators commented on src/mscorlib/src/System/Runtime/Intrinsics/X86/Sse2.cs:
```diff
@@ -14,858 +14,1080 @@ namespace System.Runtime.Intrinsics.X86
+        ///   PACKUSWB xmm, xmm/m128
         /// </summary>
         public static Vector128<byte> PackUnsignedSaturate(Vector128<short> left, Vector128<short> right) => PackUnsignedSaturate(left, right);
 
```
It seems that `void _mm_pause (void)` or `PAUSE` intrinsic is missing

tannergooding replied:
Explicitly avoiding this one as it could have undesired impacts on threading/GC, etc.

tannergooding replied:
Needs a separate, in depth discussion, to determine if it is okay to expose.

jkotas replied:
pause is already exposed as `Thread.SpinWait()`. It is not inlined by the JIT today, but that can be fixed... .

It is incredibly hard to use this instruction directly because of it has very different performance characteristics on different process (https://github.com/dotnet/coreclr/issues/13388).

4creators replied:
@tannergooding @fiigii - `void _mm_pause (void)` presence is not controlled by any CPUID flag and is available since Pentium 4 there is no reson to exclude it - anyway I remember that in corefx repo there were several issues with correctly tuning spin locks and this instruction could be a perfect solution to this problems

jkotas replied:
You have to always wrap this instruction into a construct like `Thread.SpinWait()` that gives you some consistency to get your spin loop tuned properly. I do not exposing this instruction directly would help with anything.

tannergooding replied:
@jkotas, if it is already "exposed" and shouldn't cause issues with threading or the GC, would it be good to expose here anyways, for completeness?

jkotas replied:
I do not think so.

4creators replied:
I think in such case problem is already "solved" by instruction being hard to use

4creators replied:
Seems that github is working today really very slow and I got out of the context very fast

fiigii replied:
> Thread.SpinWait() that gives you some consistency to get your spin loop tuned properly
> has very different performance characteristics on different processor models 

Right, we already did lots of work to tune the spin.

4creators commented on src/mscorlib/src/System/Runtime/Intrinsics/X86/Sse2.cs:
```diff
@@ -14,858 +14,1080 @@ namespace System.Runtime.Intrinsics.X86
+        ///   HELPER
         /// </summary>
-        public static Vector128<double> SetScalar(double value) => SetScalar(value);
+        public static Vector128<double> SetScalarVector128(double value) => SetScalarVector128(value);
```
naming: I will stop commenting on order in naming convention

4creators commented on src/mscorlib/src/System/Runtime/Intrinsics/X86/Sse2.cs:
```diff
@@ -874,536 +1096,692 @@ public static class Sse2
+        ///   PSHUFLW xmm, xmm/m128, imm8
         /// </summary>
         public static Vector128<ushort> ShuffleLow(Vector128<ushort> value, byte control) => ShuffleLow(value, control);
 
```
Functions are not in alphabetical order - perhaps it would be easier to work with them if they will be sorted

tannergooding replied:
That can be dealt with by a separate PR.

fiigii commented on src/mscorlib/src/System/Runtime/Intrinsics/X86/Sse.cs:
```diff
@@ -17,372 +17,489 @@ public static class Sse
+        /// void _mm_prefetch(char* p, int i)
+        ///   PREFETCHT1 m8
+        /// </summary>
+        public static unsafe void Prefetch1(void* address) => Prefetch1(address);
```
Shall we update other Load* intrinsics to `void*`?
For example,
```csharp
//Avx
        public static unsafe Vector256<sbyte> LoadVector256(sbyte* address) => LoadVector256(address);
        public static unsafe Vector256<byte> LoadVector256(byte* address) => LoadVector256(address);
        public static unsafe Vector256<short> LoadVector256(short* address) => LoadVector256(address);
        public static unsafe Vector256<ushort> LoadVector256(ushort* address) => LoadVector256(address);
        public static unsafe Vector256<int> LoadVector256(int* address) => LoadVector256(address);
        public static unsafe Vector256<uint> LoadVector256(uint* address) => LoadVector256(address);
        public static unsafe Vector256<long> LoadVector256(long* address) => LoadVector256(address);
        public static unsafe Vector256<ulong> LoadVector256(ulong* address) => LoadVector256(address);
        public static unsafe Vector256<float> LoadVector256(float* address) => LoadVector256(address);
        public static unsafe Vector256<double> LoadVector256(double* address) => LoadVector256(address);
```
=>
```csharp
public static unsafe Vector256<T> LoadVector256<T>(void* address) => LoadVector256<T>(address);
```

cc @jkotas @CarolEidt

4creators replied:
In this case we load into vector of predefined type and enforcing pointer to be of the same type seems correct

fiigii replied:
Agree, thanks!

tannergooding commented on src/mscorlib/src/System/Runtime/Intrinsics/X86/Avx.PlatformNotSupported.cs:
```diff
@@ -17,991 +17,1242 @@ public static class Avx
+        ///   VBROADCASTSS xmm, m32
         /// </summary>
-        public static Vector128<float> BroadcastElementToVector128(ref float source) { throw new PlatformNotSupportedException(); }
+        public static unsafe Vector128<float> BroadcastElementToVector128(float* source) { throw new PlatformNotSupportedException(); }
```
@fiigii, should these be called `BroadcastScalarToVector128`?

tannergooding replied:
(for consistency with the other operations that take a scalar.

4creators replied:
IMO in this case we have prior art in the form of established Intel docs which use term `element` and programmers IMHO probably got used to read those instructions in this way - at least this is my experience

CarolEidt replied:
That would be more consistent, I think

tannergooding replied:
Will update.

fiigii replied:
Agree.

fiigii replied:
And Avx2 counterparts should be updated as well.

tannergooding commented:
Put up the CoreFX side PR: https://github.com/dotnet/corefx/pull/26668

It is much easier to see just the API changes via it (where-as this one also contains the doc comment updates).

tannergooding commented:
@4creators, looking at the names, it actually looks to be fairly consistent already.

For example, `LoadAlignedVector128NonTemporal`:
* `Load` is the action
* `Aligned Vector128` is the type
* `Non Temporal` is the modifier

For `LoadScalarVector128`
* `Load` is the action
* `Scalar Vector128` is the type

4creators commented:
> looking at the names, it actually looks to be fairly consistent already

The point is that from my perspective and experience Vector128 represents mostly `xmm` register and action is `load aligned` or `load unaligned` which we don't use - it is personal perspective on both sides so perhaps we should ask API designers?

tannergooding commented:
> erspective and experience Vector128 represents mostly xmm register and action is load aligned or load unaligned

I think it is also consistent from that perspective, except for a couple (like `NonTemporal`, where it should be `LoadAlignedNonTemporalVector128`, from that perspective)

tannergooding commented:
@CarolEidt, @fiigii. Do you have a preference on whether it should be `LoadAlignedNonTemporalVector128`, or `LoadAlignedVector128NonTemporal` (we are currently the latter).

fiigii commented:
I prefer `LoadAlignedVector128NonTemporal`.

tannergooding commented on src/mscorlib/src/System/Runtime/Intrinsics/X86/Sse.cs:
```diff
@@ -17,372 +17,489 @@ public static class Sse
         /// __m128 _mm_cvtsi32_ss (__m128 a, int b)
+        ///   CVTSI2SS xmm, reg/m32
         /// </summary>
         public static Vector128<float> ConvertToVector128SingleScalar(Vector128<float> upper, int value) => ConvertToVector128SingleScalar(upper, value);
```
`ConvertScalarToVector128Single` might be a better name. Thoughts?

fiigii replied:
Agree.

tannergooding commented:
Ok. I think all feedback and questions have once again been answered.

Both this and the CoreFX PR (https://github.com/dotnet/corefx/pull/26668) have been updated to be in sync.

I think all missing APIs that we want to expose have been exposed. Any that aren't exposed have been discussed and agreed upon.

I think all wanted renames have now happend (except for potentially https://github.com/dotnet/coreclr/pull/15923#issuecomment-361460575, which I would like to hear from @CarolEidt and @eerhardt on).

tannergooding commented:
test this please

CarolEidt commented:
> I think it is also consistent from that perspective, except for a couple (like NonTemporal, where it should be LoadAlignedNonTemporalVector128, from that perspective)

I'm not sure I understand what perspective you're talking about. I thought that we were moving toward "Aligned Vector128" as the type, and "NonTemporal" as the modifier, and action + type + modifier it would be `LoadAlignedVector128NonTemporal`, which I think also reads better than `LoadAlignedNonTemporalVector128`

tannergooding commented:
@CarolEidt. @4creators was reading it differently, in that `LoadAligned` was the action and `Vector128` was the type, where I was viewing it as `Load` being the action and `AlignedVector128` was the type. (his perspective vs mine).

For the most part, our two perspectives actually line up, regardless of how you view the action vs type. The place they don't is for `NonTemporal`, where you could view `LoadAlignedNonTemporal` as the action and `Vector128` as the type (from @4creators perspective) or you could view it as `Load` (action), `AlignedVector128` (type), and `NonTemporal` (modifier).

This is also, actually, the only instruction I saw that has a "modifier", so it might make sense to change it to be `LoadAlignedNonTemporalVector128` (we originally called it just `LoadAlignedNonTemporal`, before we added the type in to disambiguate).

CarolEidt commented:
@tannergooding - I see now, thanks. I don't have a strong opinion either way, but if I were to be pedantic about it, I would say that alignment is very much a property of the object being loaded, while `NonTemporal` is the *way* in which the same object may or may not be loaded. So it seems to me that the `Aligned` and `Vector128` should bind more closely than the `NonTemporal`. If anything the `NonTemporal` would be part of the action (if not a modifier) and therefore: `LoadNonTemporalAlignedVector128`. That is unfortunately a bit odd because if I'm looking for aligned loads I might specifically be looking for `LoadAligned`, which leads me again to the "modifier" approach.

But, as I say, I don't have a strong opinion.

tannergooding commented:
> That is unfortunately a bit odd because if I'm looking for aligned loads I might specifically be looking for LoadAligned, which leads me again to the "modifier" approach.

I agree. I'll stick with `LoadAlignedVector128NonTemporal`, which is what @fiigii voted for as well.

That should be all the issues and this should be good for final review (assuming no other issues crop up 😄).

benaadams commented:
@tannergooding why is `PAUSE` Not Exposed (explicit)?

`Thread.SpinWait(1)` is a bit of a heavyweight way to get there

tannergooding commented:
@benaadams, I believe most of the conversation was captured here: https://github.com/dotnet/coreclr/issues/17818#issuecomment-392617183. Also see the comment chain here: https://github.com/dotnet/coreclr/pull/15923#discussion_r164619337
